File: src/ros2_bridge/mappers/control_msgs/motion_primitive.rs
//! Typed mapper for `control_msgs/msg/MotionPrimitive`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn motion_primitive_to_bus(msg: ros_env::control_msgs::msg::MotionPrimitive) -> crate::control_msgs::msg::v1::MotionPrimitive {
    crate::control_msgs::msg::v1::MotionPrimitive {
        r#type: msg.type_,
        blend_radius: msg.blend_radius,
        additional_arguments: msg.additional_arguments.into_iter().map(crate::ros2_bridge::mappers::control_msgs::motion_argument::motion_argument_to_bus).collect(),
        poses: msg.poses.into_iter().map(crate::ros2_bridge::mappers::geometry_msgs::pose_stamped::pose_stamped_to_bus).collect(),
        joint_positions: crate::ros2_bridge::mappers::convert::f64_seq(msg.joint_positions),
    }
}

pub(crate) fn motion_primitive_to_ros(bus: crate::control_msgs::msg::v1::MotionPrimitive) -> ros_env::control_msgs::msg::MotionPrimitive {
    ros_env::control_msgs::msg::MotionPrimitive {
        type_: bus.r#type,
        blend_radius: bus.blend_radius,
        additional_arguments: bus.additional_arguments.into_iter().map(crate::ros2_bridge::mappers::control_msgs::motion_argument::motion_argument_to_ros).collect(),
        poses: bus.poses.into_iter().map(crate::ros2_bridge::mappers::geometry_msgs::pose_stamped::pose_stamped_to_ros).collect(),
        joint_positions: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.joint_positions),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct ControlMsgsMotionPrimitiveMapper;

impl TypedTopicMapper for ControlMsgsMotionPrimitiveMapper {
    type Ros = ros_env::control_msgs::msg::MotionPrimitive;
    type Bus = crate::control_msgs::msg::v1::MotionPrimitive;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(motion_primitive_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(motion_primitive_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/control_msgs/motion_primitive.py
"""Generated mapper for `control_msgs/msg/MotionPrimitive`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.control_msgs.motion_argument import motion_argument_to_bus, motion_argument_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.pose_stamped import pose_stamped_to_bus, pose_stamped_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.control_msgs.msg.v1 import MotionPrimitive as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def motion_primitive_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.type = msg.type
    bus.blend_radius = msg.blend_radius
    bus.additional_arguments.extend([motion_argument_to_bus(x) for x in msg.additional_arguments])
    bus.poses.extend([pose_stamped_to_bus(x) for x in msg.poses])
    bus.joint_positions.extend(list(msg.joint_positions))
    return bus


def motion_primitive_to_ros(bus):
    from control_msgs.msg import MotionPrimitive as RosMsg

    out = RosMsg()
    out.type = bus.type
    out.blend_radius = bus.blend_radius
    out.additional_arguments = [motion_argument_to_ros(x) for x in bus.additional_arguments]
    out.poses = [pose_stamped_to_ros(x) for x in bus.poses]
    out.joint_positions = list(bus.joint_positions)
    return out


class ControlMsgsMotionPrimitiveMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from control_msgs.msg import MotionPrimitive as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return motion_primitive_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return motion_primitive_to_ros(bus)
